File: planet_species_finder.py
# Using data provided from the
#       Star Wars API (https://swapi.dev/)

# produce a
#       list

# of all possible
#       species

# for a given
#       planet

# Planet Name --> Planet URL

import json
from urllib.request import urlopen

api_key = ''  # No authentication is required to query and get data from SW-API.
planet_url = 'https://swapi.dev/api/planets/'


def api_call(url: str) -> dict:
    json_url = urlopen(url)
    return json.loads(json_url.read())  # converts json to dict


def convert_to_people(resident_list: list) -> list:
    species_set = set()
    for resident in resident_list:
        species_set |= {*api_call(url=resident).get('species')}
    return [api_call(url=species).get('name') for species in species_set]


def read_planet(planet_name: str, request_url: str = planet_url) -> dict:
    planets_dict = api_call(url=request_url)
    for planet in planets_dict.get("results"):
        if planet.get("name") == planet_name:  # if you find the planet
            return planet
    # otherwise, cycle to the next page
    return read_planet(planet_name=planet_name, request_url=next_page) if (next_page := planets_dict.get('next')) else next_page


def main(user_input: str) -> list:
    if planet_dict := read_planet(planet_name=user_input):  # check if planet exists
        return convert_to_people(resident_list=planet_dict.get('residents'))  # check if planet exists
    return ["Planet not found"]


def tester():
    # duplicate species - Tatooine
    assert main("Tatooine") == ['Droid'], "Should be {'Droid'}"
    # no species - Tholoth
    assert main("Tholoth") == [], "Should be {''}"
    # fake planet - Zatooine
    assert main("Zatooine") == ["Planet not found"], "Should be \"Planet not found\""


if __name__ == '__main__':
    tester()
    print("Everything passed")

File: planet_species_finder.js
// Using data provided from the Star Wars API(https: //swapi.dev/)
// produce a list of all possible speciesfor a given planet

// Query Path:
// Planet Name --> Planet URL --> List of Resident's URLs --> List of Species

const fetch = require("node-fetch");


const api_key = '' //No authentication is required to query and get data from SW - API.
const planet_url = 'https://swapi.dev/api/planets/'


async function api_call(url) {
    const response = await fetch(url);
    return await response.json();
}

async function convert_to_people(resident_list) {
    var species_set = []
    for (const resident of resident_list) {
        var species = await api_call(resident)
        species_set.push(species.species)
    }

    var species_arr = []
    for (const species_url of new Set(species_set.flat())) {
        var species = await api_call(species_url)
        species_arr.push(species.name);
    }
    return species_arr
}

async function read_planet(planet_name, request_url) {
    const planet_list = await api_call(request_url)

    for (const planet of planet_list.results)
        if (planet.name == planet_name) //if you find the planet
            return planet.residents

    if (next_page = planet_list.next) // otherwise, cycle to the next page
        return await read_planet(planet_name, next_page)
}

async function main(user_input) {
    if (a_val = await read_planet(user_input, planet_url)) //check if planet exists
        return await convert_to_people(a_val) //return all species on that planet
    return ["Planet not found"] //error statement
}

async function tester() {
    //duplicate species - Tatooine
    console.assert(JSON.stringify(await main('Tatooine')) === JSON.stringify(['Droid']), 'Duplicate Species assert failed');
    //no species - Tholoth
    console.assert(JSON.stringify(await main('Tholoth')) === JSON.stringify([]), 'No Species assert failed');
    //fake planet - Zatooine
    console.assert(JSON.stringify(await main('Zatooine')) === JSON.stringify(['Planet not found']), 'Fake Planet assert failed');
    console.log("Everything passed!")
}

if (process.mainModule.filename === __filename) {
    tester()
}